# Project: Investigate a Dataset (TMDb movie Database)

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> In this section of the report, I'll provide a brief introduction to the dataset I've selected for analysis. At the end of this section, I wil describe the questions that I plan on exploring over the course of the report.
> #### Introduction to dataset:
> I will be using TMDB movie dataset, This data set contains information about 10,000 movies collected from The Movie Database (TMDb), including user ratings and revenue.

In [45]:
# Use this cell to set up import statements for all of the packages that you
#   plan to use.
#importing important files 
import numpy as np
import pandas as pd
import csv
from datetime import datetime
import matplotlib.pyplot as plt
# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html
%matplotlib inline

> Let's load the data and check some rows from the dataset to identify the questions that can be answered:

In [46]:
# Reading a csv file and storing the dataset in pandas dataframe variable tmdb_data
tmdb_data = pd.read_csv('Resources/tmdb-movies.csv')
# Printing the first five rows of the dataset
tmdb_data.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline      ...       \
0              The park is open.      ...        
1             What a Lovely Day.      ...        
2     One Choice Can Destroy You      ...        
3  Every generation has a story.      ...        
4            Vengeance Hits Home      ...        

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...      3/18/15       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot     12/15/15       5292   
4  Universal Pictures|Original Film|Media Rights ...       4/1/15       2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09  

[5 rows x 21 columns]

> #### Questions that can be answered by looking at the datasets are:
> ##### Some general questions that can be answered are:
> <ol>
> <li>Which movie had the highest and lowest profit?</li>
> <li>Which movie had the greatest and least runtime?</li>
> <li>What is the average runtime of all movies?</li>
> <li>Which movie had the highest and lowest budget?</li>
> <li>Which movie had the highest and lowest revenue?</li>
> </ol>
>
> ##### Some questions that can be answered based on the profits of movie:
> <ol>
> <li>What is the average budget of the movie?</li>
> <li>What is the average revenue of the movie?</li>
> <li>What is the average runtime of the movie?</li>
> <li>Which are the successfull genres?</li>
> <li>Which are the most frequent cast involved?</li>
> </ol>

<a id='wrangling'></a>
## Data Wrangling

>In this section of the report, I will check for cleanliness, and then trim and clean my dataset for analysis.
>
> #### Observations from above dataset are:
> <ol>
> <li>The dataset has not provided the currency for columns we will be dealing with hence we will assume it is in dollars.</li>
> <li>Even the vote count is not same for all the movies and hence this affects the vote average column.</li>
> </ol>


### General Properties

> Let's check the dataset and see what cleaning does it requires.

In [47]:
# Perform operations to inspect data types and look for instances of missing or possibly errant data.
# Printing the first five rows of the dataset
tmdb_data.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline      ...       \
0              The park is open.      ...        
1             What a Lovely Day.      ...        
2     One Choice Can Destroy You      ...        
3  Every generation has a story.      ...        
4            Vengeance Hits Home      ...        

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...      3/18/15       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot     12/15/15       5292   
4  Universal Pictures|Original Film|Media Rights ...       4/1/15       2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09  

[5 rows x 21 columns]

In [48]:
# lets us check some statistics of the data
tmdb_data.describe()

id    popularity        budget       revenue       runtime  \
count   10866.000000  10866.000000  1.086600e+04  1.086600e+04  10866.000000   
mean    66064.177434      0.646441  1.462570e+07  3.982332e+07    102.070863   
std     92130.136561      1.000185  3.091321e+07  1.170035e+08     31.381405   
min         5.000000      0.000065  0.000000e+00  0.000000e+00      0.000000   
25%     10596.250000      0.207583  0.000000e+00  0.000000e+00     90.000000   
50%     20669.000000      0.383856  0.000000e+00  0.000000e+00     99.000000   
75%     75610.000000      0.713817  1.500000e+07  2.400000e+07    111.000000   
max    417859.000000     32.985763  4.250000e+08  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10866.000000  10866.000000  10866.000000  1.086600e+04  1.086600e+04  
mean     217.389748      5.974922   2001.322658  1.755104e+07  5.136436e+07  
std      575.619058      0.935142     12.812941  3.430616e+07  1.446325e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      145.750000      6.600000   2011.000000  2.085325e+07  3.369710e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

In [49]:
# Let us check infomation on datatypes of columns and missing values.
tmdb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
id                      10866 non-null int64
imdb_id                 10856 non-null object
popularity              10866 non-null float64
budget                  10866 non-null int64
revenue                 10866 non-null int64
original_title          10866 non-null object
cast                    10790 non-null object
homepage                2936 non-null object
director                10822 non-null object
tagline                 8042 non-null object
keywords                9373 non-null object
overview                10862 non-null object
runtime                 10866 non-null int64
genres                  10843 non-null object
production_companies    9836 non-null object
release_date            10866 non-null object
vote_count              10866 non-null int64
vote_average            10866 non-null float64
release_year            10866 non-null int64
budget_adj              1

In [50]:
# Calculate number of duplicated rows.
sum(tmdb_data.duplicated())

1

> #### Cleaning that needs to be prformed by looking at above data:
> <ol>
> <li>First, remove columns such as 'id', 'imdb_id', 'popularity', 'budget_adj', 'revenue_adj', 'homepage', 'keywords', 'director', 'tagline', 'overview', 'production_companies', 'vote_count' and 'vote_average'.</li>
> <li>Second, Lets delete the one duplicated row that we have in our dataset.</li>
> <li>Third, There are lots of movies where the budget or revenue have a value of '0' which means that the values of those movies has not been recorded. So we need to discard this rows, since we cannot calculate profit of such movies</li>
> <li>Fourth, The *'release_date'* column must be converted into date format.</li>
> <li>Fifth, Convert budget and revenue column to int datatype.</li>
> <li>Sixth, Replace runtime value of 0 to NAN, Since it will affect the result.</li>
> </ol>

### Data Cleaning (Lets us perform all the steps that are discussed above for cleaning)

> First, remove columns such as 'id', 'imdb_id', 'popularity', 'budget_adj', 'revenue_adj', 'homepage', 'keywords', 'director', 'tagline', 'overview', 'production_companies', 'vote_count' and 'vote_average'.

In [51]:
# Columns that needs to be deleted
deleted_columns = [ 'id', 'imdb_id', 'popularity', 'budget_adj', 'revenue_adj', 'homepage', 'keywords', 'director', 'tagline', 'overview', 'production_companies', 'vote_count', 'vote_average']
# Drop the columns from the database
tmdb_data.drop(deleted_columns, axis=1, inplace=True)
# Lets look at the new dataset
tmdb_data.head()

budget     revenue                original_title  \
0  150000000  1513528810                Jurassic World   
1  150000000   378436354            Mad Max: Fury Road   
2  110000000   295238201                     Insurgent   
3  200000000  2068178225  Star Wars: The Force Awakens   
4  190000000  1506249360                     Furious 7   

                                                cast  runtime  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...      124   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...      120   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...      119   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...      136   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...      137   

                                      genres release_date  release_year  
0  Action|Adventure|Science Fiction|Thriller       6/9/15          2015  
1  Action|Adventure|Science Fiction|Thriller      5/13/15          2015  
2         Adventure|Science Fiction|Thriller      3/18/15          2015  
3   Action|Adventure|Science Fiction|Fantasy     12/15/15          2015  
4                      Action|Crime|Thriller       4/1/15          2015

> Let's see the number of entries in our dataset now.

In [61]:
# Store rows and columns using shape function.
rows, col = tmdb_data.shape
#since rows includes count of a header, we need to remove its count.
print('We have {} total rows and {} columns.'.format(rows-1, col))

We have 3853 total rows and 8 columns.


> Second, Lets delete the one duplicated row that we have in our dataset.

In [62]:
# Drop duplicate rows but keep the first one
tmdb_data.drop_duplicates(keep = 'first', inplace = True)
# Store rows and columns using shape function.
rows, col = tmdb_data.shape
print('Now we have {} total rows and {} columns.'.format(rows-1, col))

Now we have 3853 total rows and 8 columns.


> Third, There are lots of movies where the budget or revenue have a value of '0' which means that the values of those movies has not been recorded. So we need to discard this rows, since we cannot calculate profit of such movies

In [63]:
# Columns that need to be checked.
columns = ['budget', 'revenue']
# Replace 0 with NAN
tmdb_data[columns] = tmdb_data[columns].replace(0, np.NaN)
# Drop rows which contains NAN
tmdb_data.dropna(subset = columns, inplace = True)
rows, col = tmdb_data.shape
print('We now have only {} rows.'.format(rows-1))

We now have only 3853 rows.


> Fourth, The *'release_date'* column must be converted into date format.

In [66]:
# Convert column release_date to DateTime
tmdb_data.release_date = pd.to_datetime(tmdb_data['release_date'])
# Lets look at the new dataset
tmdb_data.head()

budget       revenue                original_title  \
0  150000000.0  1.513529e+09                Jurassic World   
1  150000000.0  3.784364e+08            Mad Max: Fury Road   
2  110000000.0  2.952382e+08                     Insurgent   
3  200000000.0  2.068178e+09  Star Wars: The Force Awakens   
4  190000000.0  1.506249e+09                     Furious 7   

                                                cast  runtime  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...      124   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...      120   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...      119   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...      136   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...      137   

                                      genres release_date  release_year  
0  Action|Adventure|Science Fiction|Thriller   2015-06-09          2015  
1  Action|Adventure|Science Fiction|Thriller   2015-05-13          2015  
2         Adventure|Science Fiction|Thriller   2015-03-18          2015  
3   Action|Adventure|Science Fiction|Fantasy   2015-12-15          2015  
4                      Action|Crime|Thriller   2015-04-01          2015

> Fifth, Convert budget and revenue column to int datatype.

In [67]:
# Columns to convert datatype of
columns = ['budget', 'revenue']
# Convert budget and revenue column to int datatype
tmdb_data[columns] = tmdb_data[columns].applymap(np.int64)
# Lets look at the new datatype
tmdb_data.dtypes

budget                     int64
revenue                    int64
original_title            object
cast                      object
runtime                    int64
genres                    object
release_date      datetime64[ns]
release_year               int64
dtype: object

> Sixth, Replace runtime value of 0 to NAN, Since it will affect the result.

In [69]:
# Replace runtime value of 0 to NAN, Since it will affect the result.
tmdb_data['runtime'] = tmdb_data['runtime'].replace(0, np.NaN)
# Check the stats of dataset
tmdb_data.describe()

budget       revenue      runtime  release_year
count  3.854000e+03  3.854000e+03  3854.000000   3854.000000
mean   3.720370e+07  1.076866e+08   109.220291   2001.261028
std    4.220822e+07  1.765393e+08    19.922820     11.282575
min    1.000000e+00  2.000000e+00    15.000000   1960.000000
25%    1.000000e+07  1.360003e+07    95.000000   1995.000000
50%    2.400000e+07  4.480000e+07   106.000000   2004.000000
75%    5.000000e+07  1.242125e+08   119.000000   2010.000000
max    4.250000e+08  2.781506e+09   338.000000   2015.000000

> We have now completed our Data Wrangling section

<a id='eda'></a>
## Exploratory Data Analysis

> We will now compute statistics and create visualizations with the goal of addressing the research questions that we posed in the Introduction section.

### Research Question 1.1 (Which movie had the highest and lowest profit?)

> So we will first add a column for profit in our dataset.

In [76]:
# To calculate profit, we need to substract the budget from the revenue.
tmdb_data['profit'] = tmdb_data['revenue'] - tmdb_data['budget']
# Lets look at the new dataset
tmdb_data.head()

budget     revenue                original_title  \
0  150000000  1513528810                Jurassic World   
1  150000000   378436354            Mad Max: Fury Road   
2  110000000   295238201                     Insurgent   
3  200000000  2068178225  Star Wars: The Force Awakens   
4  190000000  1506249360                     Furious 7   

                                                cast  runtime  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...      124   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...      120   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...      119   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...      136   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...      137   

                                      genres release_date  release_year  \
0  Action|Adventure|Science Fiction|Thriller   2015-06-09          2015   
1  Action|Adventure|Science Fiction|Thriller   2015-05-13          2015   
2         Adventure|Science Fiction|Thriller   2015-03-18          2015   
3   Action|Adventure|Science Fiction|Fantasy   2015-12-15          2015   
4                      Action|Crime|Thriller   2015-04-01          2015   

       profit  
0  1363528810  
1   228436354  
2   185238201  
3  1868178225  
4  1316249360

In [79]:
# Movie with highest profit
tmdb_data.loc[tmdb_data['profit'].idxmax()]

budget                                                    237000000
revenue                                                  2781505847
original_title                                               Avatar
cast              Sam Worthington|Zoe Saldana|Sigourney Weaver|S...
runtime                                                         162
genres                     Action|Adventure|Fantasy|Science Fiction
release_date                                    2009-12-10 00:00:00
release_year                                                   2009
profit                                                   2544505847
Name: 1386, dtype: object

In [80]:
# Movie with lowest profit
tmdb_data.loc[tmdb_data['profit'].idxmin()]

budget                                                    425000000
revenue                                                    11087569
original_title                                    The Warrior's Way
cast              Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...
runtime                                                         100
genres                    Adventure|Fantasy|Action|Western|Thriller
release_date                                    2010-12-02 00:00:00
release_year                                                   2010
profit                                                   -413912431
Name: 2244, dtype: object

> #### Which movie had the highest and lowest profit?
> Highest :: Avatar with profit of 2544505847
>
> Lowest :: The Warrior's Way with profit of -413912431

### Research Question 1.2  (Which movie had the greatest and least runtime?)

### Research Question 1.3  (What is the average runtime of all movies?)

### Research Question 1.4  (Which movie had the highest and lowest budget?)

### Research Question 1.5  (Which movie had the highest and lowest revenue?)

### Research Question 2.1  (What is the average budget of the movie?)

### Research Question 2.2  (What is the average revenue of the movie?)

### Research Question 2.3  (What is the average runtime of the movie?)

### Research Question 2.4  (Which are the successfull genres?)

### Research Question 2.5  (Which are the most frequent cast involved?)

<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work, you should save a copy of the report in HTML or PDF form via the **File** > **Download as** submenu. Before exporting your report, check over it to make sure that the flow of the report is complete. You should probably remove all of the "Tip" quotes like this one so that the presentation is as tidy as possible. Congratulations!